rancher/rancher
# [v2.9-next1] Use system-default-registry when running cattle-cleanup job

## Issue: https://github.com/rancher/rancher/issues/40068

## Problem


When deleting an imported cluster from Rancher, a `cattle-cleanup` job is spawned. This job is responsible for removing rancher specific CRDs and namespaces that are no longer needed. The image used for this job does not take the system-default-registry setting into consideration. This prevents Rancher from running this job in environments where there is no access to dockerhub. 
 
## Solution
Use the `settings.SystemDefaultRegistry` setting to properly prefix the image used in the `cattle-cleanup` job. We do not support cluster level registries for imported clusters. 
 
## Testing
+ Create a pull-through private registry on a cloud provider (I used DO)
    + To create a pull through registry for testing, do the following:
 1. create a `registry` directory, and a `registry/data` directory
 2. In the `registry` directory, create a `docker-compose.yaml` file with the below content, and run `docker-compose up -d`

<details>
<summary> 
Pull Through Cache Compose File 
</summary>

```yaml
version: '3'
services:
  registry-docker.io:
    image: registry:2
    environment:
      - REGISTRY_PROXY_REMOTEURL=https://registry-1.docker.io
      - REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY=/data
    restart: always
    container_name: registry-docker.io
    ports:
      - "5000:5000"
    volumes:
      - ./data:/data
``` 
</details>

 3. Use `ngrok` (or `sslip`) to create a tunnel to the pull through registry running on port `5000`
 4. Pass the generated url to the Rancher `system-default-registry` setting 

+ After creating a testing registry, manually create a single node all roles rke2 cluster using the [quick start](https://docs.rke2.io/install/quickstart) guide

+ Create an imported cluster in Rancher, register the rke2 cluster using `kubectl`, then wait for the cluster to come online
+ Delete the cluster from Rancher 
+ using `kubectl` confirm that the `cattle-cleanup` job is pulling the `rancher-agent` image from the configured private registry. This can be done using `kubectl get events`, or `kubectl describe pod` while the job is running

## Engineering Testing
### Manual Testing
I've done the above and confirmed the system default registry is used when pulling the image 

```
12m                 Normal    SuccessfulCreate                 Job/cattle-cleanup-bfkx5                  Created pod: cattle-cleanup-bfkx5-rrpr9
12m                 Normal    Scheduled                        Pod/cattle-cleanup-bfkx5-rrpr9            Successfully assigned default/cattle-cleanup-bfkx5-rrpr9 to haffel-node
12m                 Warning   OwnerRefInvalidNamespace         ClusterRole/cattle-cleanup-qhdsx          ownerRef [batch/v1/Job, namespace: , name: cattle-cleanup-bfkx5, uid: 3d728f69-a9c7-44fe-8c17-93d3dcc0c0d5] does not exist in namespace ""
12m                 Normal    Pulling                          Pod/cattle-cleanup-bfkx5-rrpr9            Pulling image "1af7d28a0be2.ngrok.app/rancher/rancher-agent:v2.9-head"
```

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

## QA Testing Considerations


### Regressions Considerations

n/a

Existing / newly added automated tests that provide evidence there are no regressions:
n/a
